File: commonLib/src/main/java/org/treebolic/AppCompatCommonUtils.java
/*
 * Copyright (c) 2019. Bernard Bou <1313ou@gmail.com>
 */

package org.treebolic;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.Resources;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.preference.PreferenceManager;

@SuppressWarnings("WeakerAccess")
public class AppCompatCommonUtils
{
	static private final String TAG = "AppCompatCommonUtils";

	@SuppressWarnings("WeakerAccess")
	static public final String PREF_THEME = "pref_theme";

	/**
	 * Set theme preference
	 *
	 * @param context  context
	 * @param themeIdx theme idx
	 */
	static public void setThemePref(@NonNull final Context context, int themeIdx)
	{
		final SharedPreferences sharedPrefs = PreferenceManager.getDefaultSharedPreferences(context);
		final SharedPreferences.Editor editor = sharedPrefs.edit();
		editor.putInt(PREF_THEME, themeIdx);
		tryCommit(editor);
	}

	/**
	 * Get theme preference
	 *
	 * @param context context
	 * @return value (null if node)
	 */
	@Nullable
	@SuppressWarnings("boxing")
	static public Integer getThemePref(@NonNull final Context context)
	{
		final SharedPreferences sharedPrefs = PreferenceManager.getDefaultSharedPreferences(context);
		if (!sharedPrefs.contains(PREF_THEME))
		{
			return null;
		}
		try
		{
			String strValue = sharedPrefs.getString(PREF_THEME, null);
			if (strValue != null)
			{
				int intValue = Integer.parseInt(strValue);
				if (intValue != 0)
				{
					final Resources resources = context.getResources();
					final String type = resources.getResourceTypeName(intValue);
					if ("style".equals(type))
					{
						Log.d(TAG, "Theme " + type + ' ' + resources.getResourceName(intValue));
						return intValue;
					}
				}
			}
		}
		catch (NumberFormatException | Resources.NotFoundException | ClassCastException ignored)
		{
		}
		sharedPrefs.edit() //
				.remove(PREF_THEME) //
				.apply();
		return null;
	}

	/**
	 * Try to commit
	 *
	 * @param editor editor editor
	 */
	@SuppressLint({"CommitPrefEdits", "ApplySharedPref"})
	static private void tryCommit(@NonNull final SharedPreferences.Editor editor)
	{
		try
		{
			editor.apply();
		}
		catch (@NonNull final AbstractMethodError ignored)
		{
			// The app injected its own pre-Gingerbread SharedPreferences.Editor implementation without an apply method.
			editor.commit();
		}
	}
}

File: commonLib/src/main/java/org/treebolic/AppCompatCommonUtils.kt
/*
 * Copyright (c) 2019-2023. Bernard Bou
 */
package org.treebolic

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.SharedPreferences
import android.content.res.Configuration
import android.content.res.Resources.NotFoundException
import android.util.Log
import android.util.TypedValue
import androidx.core.content.edit
import androidx.core.view.WindowInsetsControllerCompat
import androidx.preference.PreferenceManager
import org.treebolic.common.R

object AppCompatCommonUtils {

    private const val TAG = "AppCompatCommonUtils"

    const val PREF_THEME: String = "pref_theme"

    /**
     * Set theme preference
     *
     * @param context  context
     * @param themeIdx theme idx
     */
    fun setThemePref(context: Context, themeIdx: Int) {
        val sharedPrefs = PreferenceManager.getDefaultSharedPreferences(context)
        val editor = sharedPrefs.edit()
        editor.putInt(PREF_THEME, themeIdx)
        tryCommit(editor)
    }

    /**
     * Get theme preference
     *
     * @param context context
     * @return value (null if node)
     */
    @JvmStatic
    fun getThemePref(context: Context): Int? {
        val sharedPrefs = PreferenceManager.getDefaultSharedPreferences(context)
        if (!sharedPrefs.contains(PREF_THEME)) {
            return null
        }
        try {
            val strValue = sharedPrefs.getString(PREF_THEME, null)
            if (strValue != null) {
                val intValue = strValue.toInt()
                if (intValue != 0) {
                    val resources = context.resources
                    val type = resources.getResourceTypeName(intValue)
                    if ("style" == type) {
                        Log.d(TAG, "Theme " + type + ' ' + resources.getResourceName(intValue))
                        return intValue
                    }
                }
            }
        } catch (_: NumberFormatException) {
        } catch (_: NotFoundException) {
        } catch (_: ClassCastException) {
        }
        sharedPrefs.edit {
            remove(PREF_THEME)
        }
        return null
    }

    /**
     * Try to commit
     *
     * @param editor editor editor
     */
    @SuppressLint("CommitPrefEdits", "ApplySharedPref")
    private fun tryCommit(editor: SharedPreferences.Editor) {
        try {
            editor.apply()
        } catch (_: AbstractMethodError) {
            editor.commit()
        }
    }

    fun isNightMode(context: Context) = (context.resources.configuration.uiMode and Configuration.UI_MODE_NIGHT_MASK) == Configuration.UI_MODE_NIGHT_YES

    fun isCurrentThemeDark(context: Context): Boolean {
        val outValue = TypedValue()
        context.theme.resolveAttribute(R.attr.isDark, outValue, true)
        return outValue.data != 0
    }

    fun isThemeDark(context: Context, themeId: Int): Boolean {
        val attrs = intArrayOf(R.attr.isDark)
        return context.obtainStyledAttributes(themeId, attrs).use { typedArray ->
            typedArray.getBoolean(0, true)
        }
    }

    fun updateBarsForTheme(activity: Activity, isDarkTheme: Boolean) {
        val controller = WindowInsetsControllerCompat(activity.window, activity.window.decorView)
        controller.isAppearanceLightStatusBars = !isDarkTheme
        controller.isAppearanceLightNavigationBars = !isDarkTheme
    }
}
